experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+      // Potential ordering issue, add commits
+      for (BlockPredecessor pred : blockPredecessors) {
+        for (OpPlacement arc : pred.arcs) {
+          placements.addCommit(arc.srcOpResult, arc.dstOp);
+          // Here, synchronizer operations will be needed in the future
+        }
+      }
+    }
+  }
+
+  // Now that new commits have been added, some of the already placed commits
+  // might be unreachable. Hence, the path to commits is marked again and
+  // unreachable commits are removed
+  markedPaths.clear();
+  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+
+  // Remove commits that cannot be reached
+  PlacementList toRemove;
+  for (OpPlacement arc : placements.getPlacements<handshake::SpecCommitOp>()) {
+    if (not markedPaths.count(arc)) {
+      toRemove.insert({arc.srcOpResult, arc.dstOp});
+    }
+  }
+  for (OpPlacement p : toRemove) {
+    placements.eraseCommit(p.srcOpResult, p.dstOp);
+  }
+}
+
+LogicalResult PlacementFinder::findCommitPositions() {
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+
+  // We need to place a commit unit before (1) an exit unit; (2) a store
+  // unit; (3) a save unit if speculative tokens can reach them.
+  llvm::DenseSet<Operation *> visited;
+  findCommitsTraversal(visited, specPos.dstOp);
+
+  // Additionally, if there are many BBs, two control-only tokens can
+  // themselves go out of order. For this reason, additional commits need
+  // to be placed in between BBs
+  findCommitsBetweenBBs();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// SaveCommit Units Finder Methods
+//===----------------------------------------------------------------------===//
+
+void PlacementFinder::findSaveCommitsTraversal(
+    llvm::DenseSet<Operation *> &visited, Operation *currOp) {
+  // End traversal if currOp is already in visited set
+  if (auto [_, isNewOp] = visited.insert(currOp); !isNewOp)
+    return;
+
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto specBB = getLogicBB(specPos.dstOp);
+
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      if (auto succOpBB = getLogicBB(succOp); !succOpBB || succOpBB != specBB) {
+        // Stop traversal if we go outside the speculation BB
+        continue;
+      } else if (placements.containsCommit(res, succOp) or
+                 placements.containsSaveCommit(res, succOp)) {
+        // End traversal on this path, as it already is cut by a commit or sc
+        continue;
+      } else if (OpPlacement specPos = placements.getSpeculatorPlacement();
+                 specPos.srcOpResult == res and specPos.dstOp == succOp) {
+        // End traversal on this path, as it already is cut by the speculator
+        continue;
+      } else if (isa<handshake::ConditionalBranchOp>(succOp)) {
+        // A SaveCommit is needed in front of the branch
+        placements.addSaveCommit(res, succOp);
```

Comment:
this seems naive also: is this how the previous approach did it?
We definitely need some kind of save-commit placement minimization eventually

Reply:
For the record, Emmet and I discussed this in a meeting and decided to address optimizations in a future pass.